File: java_benchmarks/src/main/java/com/cmpe220/benchmark/AbstractBenchmark.java
package com.cmpe220.benchmark;

abstract public class AbstractBenchmark {

    protected String category;
    protected String description;

    /**
    * Operations before the actual query
    * such as connection, initializations, etc.
    */
    protected void setupQuery() {

    }

    /**
    * Operations after the query
    */
    protected void endQuery() {

    }

    /**
    * Run the query to benchmark
    * Make sure that the query is synchronous!
    */
    abstract protected void runQuery();

    /**
    * @return milliseconds that the query took
    */
    public long benchmark() {
        setupQuery();

        long startTime = System.currentTimeMillis();
        runQuery();
        long endTime = System.currentTimeMillis();

        endQuery();
        return endTime - startTime;
    }

    public String getCategory() {
        return category;
    }

    public String getDescription() {
        return description;
    }
}

File: python_benchmarks/AbstractBenchmark.py

from abc import ABC, abstractmethod
from time import time_ns


class AbstractBenchmark(ABC):

    def __init__(self):
        self.category = ''
        self.description = ''

    def startQuery(self):
        '''
        Operations before the actual query
        e.g. connection, initializations, etc.
        '''
        pass

    def endQuery(self):
        '''
        Operations after the query
        e.g. cleanup, shutdown, etc.
        '''
        pass

    @abstractmethod
    def runQuery(self):
        '''
        Operations before the actual query
        e.g. connection, initializations, etc.
        '''
        pass

    def benchmark(self):
        '''
        :return: milliseconds that the query took
        '''
        self.setupQuery()

        start_time = time_ns()
        self.runQuery()
        end_time = time_ns()

        return (end_time - start_time) / 1000000


    def getCategory(self):
        return self.category


    def getDescription(self):
        return self.description
